use the standard syntax to load JavaScript dependencies on logistrati…

zubair-arbi opened the pull request:
…on js files

ECOM-2044
@awais786 @aamir-khan @ahsan-ul-haq @tasawernawaz 

* Use RequireJS to manage file dependencies for logistration view
* Add main view `logistration_factory.js` as main factory class on logistration page `login_and_register.html`
* Add `logistration_factory` in `modules` list of `build.js` so that RequireJS can optimize it
* Add js tests file `logistration_factory_spec` for testing the `logistration_factory`
* Add utils file `xmodule_js/common_static/js/utils/edx.utils.validate.js` in main test config file `js_test.yml`
* Add new variable `data` in the response `context` for method `login_and_registration_form`, which is then passed directly to `logistration_factory` as initial configuration
* Fix failing python unit tests of `student_account` and `third_party_auth` views

Sandbox: http://zubair-arbi.sandbox.edx.org/ (updated)

zubair-arbi commented:
@will I have refactored the views and models for logistration so that requirejs optimizer can be used for relevant views as mentioned in [this document](https://openedx.atlassian.net/wiki/display/TNL/How+to+add+a+new+page+to+LMS+or+Studio#HowtoaddanewpagetoLMSorStudio-requirejs-text).
You can test the logistration factory on sandbox: http://zubair-arbi.sandbox.edx.org/register

zubair-arbi commented:
@AlasdairSwan fyi

andy-armstrong commented on lms/static/js/student_account/logistration_factory.js:
```diff
@@ -0,0 +1,21 @@
+
+        var container = $('#login-and-registration-container');
+
+        return function () {
```
Nit: my preference is to pass a dict of options to the factory, rather than have it depend upon data attributes that happen to be set in the environment. This makes the factory easier to test in isolation, and also makes the code easier to understand. For one example, see how the Teams Mako template interacts with its factory:

https://github.com/edx/edx-platform/blob/master/lms/djangoapps/teams/static/teams/js/teams_tab_factory.js
https://github.com/edx/edx-platform/blob/master/lms/djangoapps/teams/templates/teams/teams.html

Some of our other views construct the configuration dict in the Python view, pass it to the Mako template, and then use ``json.dumps`` to write it out as a JavaScript object. For example:

https://github.com/edx/edx-platform/blob/master/lms/static/js/student_profile/views/learner_profile_factory.js
https://github.com/edx/edx-platform/blob/master/lms/templates/student_profile/learner_profile.html
https://github.com/edx/edx-platform/blob/master/lms/djangoapps/student_profile/views.py#L41

My preference is for the latter because it keeps all the configuration in one place in the Python code that can be easily tested.

andy-armstrong commented:
FYI @cahrens

andy-armstrong commented:
jenkins run js

zubair-arbi commented:
@andy-armstrong I have addressed your comments. Added a field `data` in response `context` of logistration view, which I pass directly to `logistration_factory` as initial configuration. Also added some basic tests for initial rendering of `logistration_factory`, I didn't added tests for all event and scenarios since those case are already being tested in the file `access_spec.js`. Refactored some additional files e.g. `edx.utils.validate` which was causing issues while using in the testing.
@AlasdairSwan @andy-armstrong @wedaly This PR is ready for final review. I am updating the sandbox, it will be available after some time (~30min from now) at url: http://zubair-arbi.sandbox.edx.org/

andy-armstrong commented on common/static/js/utils/edx.utils.validate.js:
```diff
@@ -1,186 +1,187 @@
-var edx = edx || {};
-
-(function( $, _, _s, gettext ) {
+;(function (define) {
```
Since you're changing this class, I'd recommend also moving it to``common/static/common/js`` which is where we want all common code to live. The idea is that this is namespaced and so it can be safely used in other contexts. I'd probably name it:

    common/static/common/js/utils/validate.js

@symbolist I'm not sure if you're aware that we now have two very similar JavaScript libraries for building client-side forms. There's the one for logistration, and then also the one used by account settings. In the long run we should refactor these into a single framework.

symbolist replied:
@andy-armstrong Thanks! I wasn't. Hopefully one day! :)

andy-armstrong commented:
@zubair-arbi This is great stuff! If you wouldn't mind moving the class I mention (and its associated spec file) that would be great.

zubair-arbi commented:
jenkins run bokchoy

zubair-arbi commented:
@andy-armstrong I wouldn't mind that, but the thing is this file `edx.utils.validate.js` already given me too much trouble. I debugged for two day's due to path conflict (which is my guess) after I added path `xmodule_js/common_static/js` in directory `src_paths` of file `js_test.yaml` to include it in tests and require js started timing out on other different js files. So I don't want to touch it again until its necessary :)

zubair-arbi commented:
jenkins run bokchoy

andy-armstrong commented:
@zubair-arbi No problem. Just go ahead and merge it as is.

zubair-arbi commented:
jenkins run bokchoy

zubair-arbi commented:
@awais786 Please review

zubair-arbi commented:
@muhhshoaib Please review this PR.

AlasdairSwan commented on common/static/js/utils/edx.utils.validate.js:
```diff
@@ -1,186 +1,187 @@
+        }
+        _.mixin( _s.exports() );
+
+        var utils = (function(){
```
Variables should be defined on the first line of a function due to hoisting.

AlasdairSwan commented on lms/static/js/spec/student_account/access_spec.js:
```diff
@@ -51,22 +49,24 @@ define([
+                    registration_form_desc: FORM_DESCRIPTION,
+                    password_reset_form_desc: FORM_DESCRIPTION
+                };
+                var logistrationElement = $('#login-and-registration-container');
```
There only needs to to be 1 var used at the top of the function, comma separated to define multiple variables.

AlasdairSwan commented on lms/static/js/spec/student_account/hinted_login_spec.js:
```diff
@@ -1,87 +1,85 @@
-        it('redirects the user to the hinted provider if the user clicks the proceed button', function() {
-            createHintedLoginView("oa2-google-oauth2");
+            it('redirects the user to the hinted provider if the user clicks the proceed button', function() {
+                createHintedLoginView("oa2-google-oauth2");
```
For consistency please update to use single quotes

AlasdairSwan commented on lms/static/js/spec/student_account/institution_login_spec.js:
```diff
@@ -1,80 +1,79 @@
+            it('displays a list of providers', function() {
+                createInstLoginView('register');
+                expect($('#institution_login-form').html()).not.toBe("");
+                var $google = $('li a:contains("Google")');
```
Variables should be declared at the top of the function.

AlasdairSwan commented on lms/static/js/spec/student_account/institution_login_spec.js:
```diff
@@ -1,80 +1,79 @@
+                expect($google).toHaveAttr(
+                    'href', '/auth/login/google-oauth2/?auth_entry=account_register'
+                );
+                var $facebook = $('li a:contains("Facebook")');
```
Variables should be declared at the top of the function.

AlasdairSwan commented on lms/static/js/spec/student_account/login_spec.js:
```diff
@@ -1,265 +1,262 @@
+                $('#login-remember').prop('checked', USER_DATA.remember);
+
+                // Create a fake click event
+                var clickEvent = $.Event('click');
```
Variables should be declared at the top of the function.

AlasdairSwan commented on lms/static/js/spec/student_account/login_spec.js:
```diff
@@ -1,265 +1,262 @@
+                // Simulate that the user is attempting to enroll in a course
+                // by setting the course_id query string param.
+                spyOn($, 'url').andCallFake(function( param ) {
+                    if (param === "?course_id") {
```
Please update to single quotes for consistency

AlasdairSwan commented on lms/static/js/spec/student_account/register_spec.js:
```diff
@@ -1,352 +1,348 @@
+                $('#register-honor_code').prop('checked', USER_DATA.honor_code);
+
+                // Create a fake click event
+                var clickEvent = $.Event('click');
```
Variables should be declared at the top of the function.

AlasdairSwan commented on lms/static/js/spec/student_account/register_spec.js:
```diff
@@ -1,352 +1,348 @@
-        it('displays third-party auth registration buttons', function() {
-            createRegisterView(this);
+                // Verify that the client sent the course ID for analytics
+                var expectedData = {course_id: COURSE_ID};
```
Variables should be declared at the top of the function.

AlasdairSwan commented on lms/static/js/student_account/logistration_factory.js:
```diff
@@ -0,0 +1,12 @@
+    define(['jquery', 'js/student_account/views/AccessView'],
+        function($, AccessView) {
+
+        return function (options) {
```
There should not be a space between function and (, it should be
```
function(options) {
```

AlasdairSwan commented on lms/static/js/student_account/models/LoginModel.js:
```diff
@@ -1,58 +1,54 @@
     'use strict';
+    define(['jquery', 'backbone', 'jquery.url'], function ($, Backbone) {
+
+        var LoginModel = Backbone.Model.extend({
```
Instead of assigning the model to a variable and returning that you can just return the model, i.e.
```
return Backbone.Model.extend({
```

AlasdairSwan commented on lms/static/js/student_account/models/PasswordResetModel.js:
```diff
@@ -1,44 +1,40 @@
-        }
+    define(['jquery', 'backbone'], function($, Backbone) {
+
+        var PasswordResetModel = Backbone.Model.extend({
```
Instead of assigning the model to a variable and returning that you can just return the model, i.e.
```
return Backbone.Model.extend({
```

AlasdairSwan commented on lms/static/js/student_account/models/RegisterModel.js:
```diff
@@ -1,61 +1,57 @@
     'use strict';
+    define(['jquery', 'backbone', 'jquery.url'], function($, Backbone) {
+
+        var RegisterModel = Backbone.Model.extend({
```
Instead of assigning the model to a variable and returning that you can just return the model, i.e.
```
return Backbone.Model.extend({
```

AlasdairSwan commented on lms/static/js/student_account/views/AccessView.js:
```diff
@@ -1,280 +1,283 @@
-        },
-
-        nextUrl: '/dashboard',
+        var AccessView = Backbone.View.extend({
```
Instead of assigning the view to a variable and returning that you can just return the viewl, i.e.
```
return Backbone.View.extend({
```

AlasdairSwan commented on lms/static/js/student_account/views/AccessView.js:
```diff
@@ -1,280 +1,283 @@
+            passwordEmailSent: function() {
+                this.element.hide( $(this.el).find('#password-reset-anchor') );
+                this.element.show( $('#login-anchor') );
+                this.element.scrollTop( $('#login-anchor') );
```
The #login-anchor jQuery element is being created twice, it would be better to make it a variable so you are just creating the object once and using it twice.

AlasdairSwan commented on lms/static/js/student_account/views/FormView.js:
```diff
@@ -1,270 +1,267 @@
-
-    edx.student.account.FormView = Backbone.View.extend({
-        tagName: 'form',
+        var FormView = Backbone.View.extend({
```
Instead of assigning the view to a variable and returning that you can just return the viewl, i.e.
```
return Backbone.View.extend({
```

AlasdairSwan commented on lms/static/js/student_account/views/HintedLoginView.js:
```diff
@@ -1,48 +1,46 @@
-        }
+    define(['jquery', 'underscore', 'backbone'], function($, _, Backbone) {
+
+        var HintedLoginView = Backbone.View.extend({
```
Instead of assigning the view to a variable and returning that you can just return the view, i.e.
```
return Backbone.View.extend({
```

AlasdairSwan commented on lms/static/js/student_account/views/InstitutionLoginView.js:
```diff
@@ -1,30 +1,28 @@
-
-    edx.student.account.InstitutionLoginView = Backbone.View.extend({
-        el: '#institution_login-form',
+        var InstitutionLoginView = Backbone.View.extend({
```
Instead of assigning the view to a variable and returning that you can just return the view, i.e.
```
return Backbone.View.extend({
```

AlasdairSwan commented on lms/static/js/student_account/views/LoginView.js:
```diff
@@ -1,131 +1,125 @@
-                    platformName: this.platformName
+    define(['jquery', 'underscore', 'js/student_account/views/FormView'], function($, _, FormView) {
+
+        var LoginView = FormView.extend({
```
Instead of assigning the view to a variable and returning that you can just return the view, i.e.
```
return FormView.extend({
```

AlasdairSwan commented on lms/static/js/student_account/views/PasswordResetView.js:
```diff
@@ -1,48 +1,45 @@
-
-    edx.student.account.PasswordResetView = edx.student.account.FormView.extend({
-        el: '#password-reset-form',
+        var PasswordResetView = FormView.extend({
```
Instead of assigning the view to a variable and returning that you can just return the view, i.e.
```
return FormView.extend({
```

AlasdairSwan commented on lms/static/js/student_account/views/RegisterView.js:
```diff
@@ -1,100 +1,98 @@
     'use strict';
+    define(['jquery', 'underscore', 'js/student_account/views/FormView'], function($, _, FormView) {
+
+        var RegisterView = FormView.extend({
```
Instead of assigning the view to a variable and returning that you can just return the view, i.e.
```
return FormView.extend({
```

AlasdairSwan commented:
Pretty minor comments but important to make I think. It may be out os scope of this PR but it would be great if the templates being used by the views were underscore templates passed into the view instead of having to render them on the page and get them by id selector

zubair-arbi commented:
jenkins run all

zubair-arbi commented:
jenkins run python

zubair-arbi commented:
jenkins run python

zubair-arbi commented:
@AlasdairSwan I have addressed all your comments. Please give a final review.

AlasdairSwan commented:
Looks good @zubair-arbi :+1:
